crates/iota-data-ingestion-core/src/reducer.rs

```diff
@@ -144,6 +146,51 @@ pub(crate) async fn reduce<W: Worker>(
                 .map_err(|_| IngestionError::Channel("unable to send worker pool progress updates to executor, receiver half closed".into()))?;
             progress_update = None;
         }
+
+        // Check for shutdown signal after progress update to ensure progress
+        // is reported even during shutdown.
+        if trigger_shutdown {
+            break;
+        }
     }
     Ok(())
 }
+
+/// Represents the outcome of a commit operation with retry logic.
+enum CommitStatus {
+    /// Commit succeeded, continue processing.
+    Success,
+    /// Processing should stop due to shutdown signal.
+    Shutdown,
+}
+
+/// Commits a batch with exponential backoff retries on failure.
+///
+/// Returns:
+/// - `CommitStatus::Success`: On successful commit.
+/// - `CommitStatus::Shutdown`: If commit failed and shutdown signal was
+///   received.
+///
+/// # Panics:
+/// If all retry attempts are exhausted within the 15-minute maximum elapsed
+/// time.
+async fn commit_with_retry<W: Worker>(
+    reducer: &dyn Reducer<W>,
+    batch: Arc<Vec<<W as Worker>::Message>>,
+    token: &CancellationToken,
+) -> CommitStatus {
+    let backoff = backoff::ExponentialBackoff::default();
+    backoff::future::retry(backoff, || async {
+        let commit_result = reducer.commit(batch.clone()).await.map_err(|err| {
+            let err = IngestionError::Reducer(format!("failed to commit batch: {err}"));
+            tracing::warn!("transient reducer commit error {err:?}");
+            backoff::Error::transient(err)
+        });
+        if commit_result.is_err() && token.is_cancelled() {
```

Comment:
Will we not hide the cancellation signal that way?
The `reduce` function is only checking for `Shutdown` return value, it's not checking the token itself.

Reply:
fixed: https://github.com/iotaledger/iota/pull/5820/commits/da658404e6778cfac7c2fc5de93126cce9b81cd0#diff-3665ba807b5937734070baa5b4c12186d05842a235eb9b5422acd2a87c4cbe63R188-R191